tests/h/traversal/annotation_test.py

```diff
@@ -1,348 +1,134 @@
-            else:
-                acl.append((security.Allow, p, Permission.Group.FLAG))
-                # Normally, the ``moderate`` permission would only be applied
-                # to the admin (creator) of a group, but this ``FakeGroup``
-                # is indeed fake. Tests in this module are merely around whether
-                # this permission is translated appropriately from a group
-                # to an annotation context (i.e. it should not be applied
-                # to private annotations)
-                acl.append((security.Allow, p, Permission.Group.MODERATE))
-        self.__acl__ = acl
-
-
-@pytest.mark.usefixtures("groupfinder_service", "links_service")
 class TestAnnotationRoot:
-    def test_it_does_not_assign_create_permission_without_authenticated_user(
-        self, set_permissions, pyramid_request
-    ):
-        set_permissions()
-
-        context = AnnotationRoot(pyramid_request)
+    def test_create_permission_requires_authenticated_user(self, root, permits):
+        assert permits(root, [security.Authenticated], Permission.Annotation.CREATE)
+        assert not permits(root, [], Permission.Annotation.CREATE)
 
-        assert not pyramid_request.has_permission(Permission.Annotation.CREATE, context)
-
-    def test_it_assigns_create_permission_to_authenticated_request(
-        self, set_permissions, pyramid_request
+    def test_annotation_lookup(
+        self,
+        root,
+        pyramid_request,
+        AnnotationContext,
+        storage,
+        groupfinder_service,
+        links_service,
     ):
-        set_permissions(
-            "acct:adminuser@foo", principals=[pyramid.security.Authenticated]
+        context = root[sentinel.annotation_id]
+
+        assert context == AnnotationContext.return_value
+        storage.fetch_annotation.assert_called_once_with(
+            pyramid_request.db, sentinel.annotation_id
+        )
+        AnnotationContext.assert_called_once_with(
+            storage.fetch_annotation.return_value, groupfinder_service, links_service
         )
 
-        context = AnnotationRoot(pyramid_request)
+    def test_failing_annotation_lookup(self, root, storage):
+        storage.fetch_annotation.return_value = None
+
+        with pytest.raises(KeyError):
+            assert root[sentinel.annotation_id]
 
-        assert pyramid_request.has_permission(Permission.Annotation.CREATE, context)
+    @pytest.fixture
+    def root(self, pyramid_request):
+        return AnnotationRoot(pyramid_request)
 
-    def test_get_item_fetches_annotation(self, pyramid_request, storage):
-        factory = AnnotationRoot(pyramid_request)
+    @pytest.fixture
+    def storage(self, patch):
+        return patch("h.traversal.annotation.storage")
 
-        factory["123"]
-        storage.fetch_annotation.assert_called_once_with(pyramid_request.db, "123")
+    @pytest.fixture
+    def AnnotationContext(self, patch):
+        return patch("h.traversal.annotation.AnnotationContext")
```

Comment:
These two patches should probably be `autouse` and module-level functions instead of methods

Reply:
The storage method is only used in this class and not the other, and the `AnnotationContext` is the class under test for the rest of the file, so we can't mock it out.

I think the rule about doing this probably needs some extra parts to it:

* This only applies when there's a single class under test in a set of tests
* This only applies when there are multiple test classes for that test

In the case where there are multiple classes under test, or there is a single test class it's probably best to keep the fixtures local to it.

I like keeping the mocks local to where they are needed for a few reasons:

* It limits the number of things in play - If you look at the end of the file and there are 10 auto-uses, but only 3 have any relevance it makes it hard to know which things matter
* It mentally decouples the tests - The tests below have nothing to do with this fixture, and implying they do complicates them unnecessarily
* It makes it a breeze to refactor tests when you want to move classes, as you don't have to sort through the fixtures to work out which ones belong to what. You can just scoop up the class and take it's fixtures with it
* A general preference for local over global